src/common/browser-adapters/browser-adapter-event-manager.ts

```diff
@@ -14,115 +13,150 @@ export interface EventDetails {
+    protected deferredEvents: DeferredEventDetails[] = [];
+    protected eventsToApplicationListenersMapping: DictionaryStringTo<ApplicationListener> = {};
+    protected eventsToAdapterListenersMapping: DictionaryStringTo<AdapterListener> = {};
+    protected handleEvent: (eventType: string) => AdapterListener =
         (eventType: string) =>
         (...eventArgs: any[]) => {
-            this.processEvent(eventType, eventArgs);
+            const processedResults = this.processEvent(eventType, eventArgs); //guard in case we return undefined for a timed out promise resoponse
+            return processedResults === null
+                ? this.deferEvent({ eventType, eventArgs })
+                : processedResults;
         };
-    constructor(private promiseFactory: PromiseFactory, private timeoutFactory: TimeoutFactory) {}
+
+    constructor(private promiseFactory: PromiseFactory, private logger: Logger) {}
 
     public registerEventToApplicationListener = (
         eventType: string,
         callback: ApplicationListener,
     ): void => {
         if (this.eventsToApplicationListenersMapping[eventType]) {
-            this.eventsToApplicationListenersMapping[eventType].push(callback);
-        } else {
-            this.eventsToApplicationListenersMapping[eventType] = [callback];
+            throw new Error(`Listener already registered for ${eventType}`);
         }
+        this.eventsToApplicationListenersMapping[eventType] = callback;
         this.processDeferredEvents();
     };
 
     public registerAdapterListenerForEvent = (
         event: Events.Event<any>,
         eventType: string,
     ): void => {
-        event.addListener(this.adapterListener(eventType));
+        const eventListener = this.handleEvent(eventType);
+        event.addListener(eventListener);
+        this.eventsToAdapterListenersMapping[eventType] = eventListener;
     };
 
+    public processEvent(eventType: string, eventArgs: any[]): Promise<any> | null {
+        const applicationListener = this.eventsToApplicationListenersMapping[eventType];
+        return applicationListener
+            ? this.forwardEventToApplicationListener(applicationListener, eventArgs)
+            : null;
+    }
+
+    public removeListener(event: Events.Event<any>, eventType: string) {
+        this.unregisterAdapterListener(event, eventType);
+        this.unregisterApplicationListenerForEvent(eventType);
+    }
+
     private processDeferredEvents(): void {
-        this.deferredEvents = this.deferredEvents.filter(
-            event => this.processEvent(event.eventType, event.eventArgs, true) === false,
-        );
+        const stillDeferred: DeferredEventDetails[] = [];
+        for (const deferredEvent of this.deferredEvents) {
+            if (this.eventsToApplicationListenersMapping[deferredEvent.eventType]) {
+                try {
+                    this.processEvent(deferredEvent.eventType, deferredEvent.eventArgs);
```

Comment:
I don't think it's quite right for this to ignore the return value of `this.processEvent`; I think this needs to set it up so that if `processEvent` returns a `Promise`, the `deferredResolution.resolveHook` gets invoked only after that promise settles, not immediately.